File: c++/Trees/UnionFind/impl.cpp

#include <map>
using namespace std;
/**
 * Union find algorithm with path compression in the find algorithms
 * What happens to the rank? Time complexity if O(ElogV) but reduces to O(1)?
 * More on this here: https://www.geeksforgeeks.org/union-by-rank-and-path-compression-in-union-find-algorithm/
 * Useful when you want to see connected aspects of a graph.
 */
class UnionFind
{

private:
    map<int, int> par;
    map<int, int> rank;
    int num_comp;

public:
    UnionFind(int n)
        : num_comp(n)
    {
        for (int i = 0; i < n; i++)
        {
            par[i] = i;
            rank[i] = 0;
        }
    }

    int find(int x)
    {
        if (par[x] != x)
        {
            par[x] = find(par[x]);
        }
        return par[x];
    }

    bool isSameComponent(int x, int y)
    {
        return (find(x) == find(y));
    }

    // Union is a reserved keyword in C++, so we use _union instead
    bool _union(int x, int y)
    {
        int xp = find(x), yp = find(y);
        if (xp == yp)
            return false;

        if (rank[xp] > rank[yp])
            par[yp] = xp;

        else if (rank[yp] > rank[xp])
            par[xp] = yp;

        else
        {
            par[yp] = xp;
            rank[xp] += 1;
        }
        num_comp--;
        return true;
    }

    int getNumComponents()
    {
        return num_comp;
    }
};

File: Python/Trees/UnionFind/impl.py


""" The following is a union find. It is useful when determining is parts of a graph are connected 
    Due to the optimiations, the time complexity for the find (the bottleneck operation) reduces to O(1).
    More on this here: https://www.geeksforgeeks.org/union-by-rank-and-path-compression-in-union-find-algorithm/"""

""" Although the rank is temporarly altered after the find funciton, it's still accurate at the time of the union """


class UnionFind:
    
    def __init__(self, n: int):
        self.par = {}
        self.rank = {}
        self.num_comp = n

        for i in range(n):
            self.par[i] = i
            self.rank[i] = 0

    def find(self, x: int) -> int:
        # here we do path compression. How does this impact the rank? It becomes arbitrary.
        if x != self.par[x]:
            self.par[x] = self.find(self.par[x])
        return self.par[x]

    def isSameComponent(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)

    def union(self, x: int, y: int) -> bool:
        xp, yp = self.find(x), self.find(y)
        if (xp == yp):
            return False

        if self.rank[xp] > self.rank[yp]:
            self.par[yp] = xp
        
        elif self.rank[xp] < self.rank[yp]:
            self.par[xp] = yp

        else:
            self.par[yp] = xp
            self.rank[xp] = 1
        self.num_comp -= 1
        return True

    def getNumComponents(self) -> int:
        return self.num_comp

